File: Python_solutions/LC739.py
# LC 739 Daily Temperatures
temperatures = [73,74,75,71,69,72,76,73]

def solution(temperatures):
    # •	Implement two set
    r = len(temperatures)
    stack = []
    days = ([0] * r)
    for l in range(r):
        # •	Compare with the last element(stack) of the stack with current element(ip)
        while stack and (temperatures[stack[-1]] < temperatures[l]):
            # •	Calculate the day count with index values
            days[stack[-1]] = ((l) - stack[-1])
            # •	Remove the index value from stack
            stack.pop()
        stack.append(l)
    return days

print(solution(temperatures))

File: MonotonicStack/C++_solution/LC739.cpp
// LC: 739 https://leetcode.com/problems/daily-temperatures/

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int r = temperatures.size();
        vector<int> days(r, 0);
        
        // Raw array on the Stack
        int stack[100000]; 
        int top = -1; // Our physical pointer
        
        for (int l = 0; l < r; l++) {
            
            // While stack has items AND current temp > temp at waiting index
            while (top >= 0 && temperatures[stack[top]] < temperatures[l]) {
                
                // Grab the waiting index
                int waiting_index = stack[top];
                
                // Calculate and store the difference in days
                days[waiting_index] = l - waiting_index;
                
                // Pop the pointer
                top--;
            }
            
            // Push current index into the waiting room
            stack[++top] = l;
        }
        
        return days;
    }
};